src/db/roomstore.ts

```diff
@@ -0,0 +1,365 @@
+                this.entriesMatrixIdCache.delete(matrixId);
+            }
+            const items: string[] = [];
+
+            if (mxIdDifferent) {
+                items.push("matrix_id = $matrixId");
+            }
+
+            if (rmIdDifferent) {
+                items.push("remote_id = $remoteId");
+            }
+
+            await this.db.Run(`UPDATE room_entries SET ${items.join(", ")} WHERE id = $id`,
+                {
+                    id: entry.id,
+                    matrixId: matrixId as string|null,
+                    remoteId: remoteId as string|null,
+                },
+            );
+        }
+
+        // Matrix room doesn't store any data.
+        if (entry.remote) {
+            await this.upsertRoom(entry.remote);
+        }
+    }
+
+    public async getEntriesByMatrixId(matrixId: string): Promise<IRoomStoreEntry[]> {
+        const cached = this.entriesMatrixIdCache.get(matrixId);
+        if (cached && cached.ts + ENTRY_CACHE_LIMETIME > Date.now()) {
+            return cached.e;
+        }
+        const entries = await this.db.All(
+            "SELECT * FROM room_entries WHERE matrix_id = $id", {id: matrixId},
+        );
+        const res: IRoomStoreEntry[] = [];
+        for (const entry of entries) {
+            let remote: RemoteStoreRoom|null = null;
+            if (entry.remote_id) {
+                const remoteId = entry.remote_id as string;
+                const row = await this.db.Get(
+                    "SELECT * FROM remote_room_data WHERE room_id = $remoteId",
+                    {remoteId},
+                );
+                if (row) {
+                    // tslint:disable-next-line no-any
+                    remote = new RemoteStoreRoom(remoteId, row as any);
+                }
+            }
+
+            res.push({
+                id: (entry.id as string),
+                matrix: new MatrixStoreRoom(matrixId),
+                remote,
+            });
+        }
+        if (res.length > 0) {
+            this.entriesMatrixIdCache.set(matrixId, {e: res, ts: Date.now()});
+        }
+        return res;
+    }
+
+    public async getEntriesByMatrixIds(matrixIds: string[]): Promise<IRoomStoreEntry[]> {
+        const mxIdMap = { };
+        matrixIds.forEach((mxId, i) => mxIdMap[i] = mxId);
+        const sql = `SELECT * FROM room_entries WHERE matrix_id IN (${matrixIds.map((_, id) => `\$${id}`).join(", ")})`;
+        const entries = await this.db.All(sql, mxIdMap);
+        const res: IRoomStoreEntry[] = [];
+        for (const entry of entries) {
+            let remote: RemoteStoreRoom|null = null;
+            const matrixId = entry.matrix_id as string || "";
+            const remoteId = entry.remote_id as string;
+            if (remoteId) {
+                const row = await this.db.Get(
+                    "SELECT * FROM remote_room_data WHERE room_id = $rid",
+                    {rid: remoteId},
+                );
+                if (row) {
```

Comment:
same here, maybe it is actually a good idea to ignore results without remote data (like you did originally)